content/posts/2021-01-fun-with-iconfiguration.md

```diff
@@ -0,0 +1,285 @@
+
+However, in the absence of equivalent services in the data centre (which we _will_ provision eventually - things like Vault for secrets and Consul for service discovery are good options here) we are stuck with using `environmentsettings.{environment}.json` as a way to specify the data centre environment :(
+
+## Using KeyVault & AppConfig
+
+Our use of both KeyVault & AppConfig is pretty simple. When an environment is provisioned in Azure using Terraform some global secrets such as SQL admin credentials and credentials to third party systems are written into KeyVault. Once that process has completed we run a bootstrap script that consumes some of those secrets and provisions our SQL accounts, shared secrets and introspects the environment to configure service endpoints. Any key/value pairs are persisted into KeyVault or AppConfig depending on the kind of value it is (secret/service metadata/etc).
+
+In the application we configure AppConfig & KeyVault providers using managed identity (hooray, no secrets in JSON!) which means we can consume `IConfiguration` in our site settings loader and dependent settings are configured appropriately.
+
+However, we quickly discovered that having individual settings, potentially composed of other settings, becomes somewhat unmaintainable. Consider a connection string to SQL:
+
+> `Server=sql.database.windows.net;Database=MyDatabase;User ID=myapp-readwrite;Password=Password123!`
+
+This contains the hostname of a server and the credentials used to connect to it. These are all discrete pieces of information, with their own, potentially disconnected, lifetimes, that are composed to form the whole connection string. That means that any time one of those pieces of information changes we must reconstruct all dependent values from its constituent pieces again. Eurgh!
+
+But, thankfully, there's a simple solution!
+
+### Substitution
+
+If we change the configuration values composed from other values to use placeholders then we can change our connection string to:
+
+> `Server=${SqlServer};Database=MyDatabase;User ID=${SqlUser};Password=${SqlPassword}`
+
+But this doesn't work out of the box! To make this happen we've invented a configuration provider that wraps other providers and allows them to use  values from elsewhere in the configuration system. We can do this by including the `StackExchange.Utils.Configuration` package and configuring our application as follows:
+
+```c#
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        WebHost.CreateDefaultBuilder(args)
+            .ConfigureAppConfiguration(
+                (hostingContext, config) =>
+                {
+                    config
+                        // environmental defaults - service endpoints, servers
+                        .AddJsonFile("globals.json")
+                        // environmental secrets - API keys, passwords
+                        .AddJsonFile("secrets.json")
```

Comment:
I assume we always deploy `.json` files and our application in lock-step? Otherwise, would it make sense to go over the `reloadOnChange: true` parameter?

Reply:
In our case - yes we do. Generally though I didn't wanna go too deep into the underlying intricacies of each provider, so left reloads out of this bit!